Introduce scenario-based testing for Byzantine faulty cases

nhoriguchi opened the pull request:
Hi,

I updated testing framework for Byzantine faulty case with scenario-based one based on discussion #169. It's still working in progress but the base is taking shape, so I feel that it's enough to start discussion for feedback. I share below some points which might interest you:

- I think that adding new public interface `NonDefaultMessageHandler` to package `minbft` is a kind of dependency-injection approach (but there might be better one),
- "selective ignorant" case [mentioned](https://github.com/hyperledger-labs/minbft/pull/169#issuecomment-617614924) in #169 could be implemented on connection layer, but in current approach it's controlled by test simulator/controller,
- the testcase `testManipulatedOperation` seems to reach consensus improperly, where the primary replica manipulates operation string in request message and the manipulated string is stored to `SimpleLedger`. So this might be one example case that one Byzantine faulty node could compromise the operation of MinBFT cluster.
- scenario interpreter is still roughly designed, and it needs to be improved for more generic cases. One known issue is that checking phase fails on assertion when waiting for multiple messages of the same type in random order.
- I have to update README.md to explain some about Byzantine faulty case.

Any comments and suggestions are welcomed!

nhoriguchi committed:
core: move utils.go to separate internal package

Some code in utils.go will be reused from another package minbft_test
later, so let's move the code into a separate package utils.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi committed:
core: make type messageHandler public

messageHandler will be used from another package minbft_test later,
so let's make it public.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi committed:
core/replica: introduce NonDefaultMessageHandler interface

Currently replica instance defines message handler functions by
generator functions hard-coded in the function New(), This
patch proposes to allow external code to inject custom generator
for these message handlers.

The background of this change is to enable more extensive tesitng,
especially in Byzantine faulty scenarios, where we try to see
whether a MinBFT replica cluster reaches consensus if there're
some faulty nodes. The details will be explained in later patches.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi committed:
core: make type makeHelloHandler public

This will be used from another package minbft_test later,
so let's make it public.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi committed:
core: move usig-ui.go code to separate internal package

usig-ui.go will be used from separate package minbft_test later, so
let's move it (and related test code) to a new package usigui.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi committed:
minbft_test: introduce basic testcases with simulated replicas

This patch introduce an infrastructure of scenario-based tesitng
with simulated replicas. This could provide flexible testing
function like Byzantine faulty cases, but this patch only starts
with simple non-faulty case to demonstrate and confirm that the
infrastructure properly works.

Simulator runs "environment network" which interacts with the
test target replica, and has characteristics like below:
- it mimics all nodes except the target one, which means that
  we are not interested only in simulator-target communication,
  not in communication between nodes within simulator,
- it's behavior is controlled by single yaml which defines
  what messages are emitted from the target replica as a
  response to the specific input messages from simlator.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi committed:
[WIP] minbft_test: introduce simple Byzantine faulty testcases

Based on the simulation-based testing infrastructure, this
patch introduces first few Byzantine faulty testcases.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

ynamiki commented on core/replica.go:
```diff
@@ -72,8 +76,14 @@ func New(id uint32, configer api.Configer, stack Stack, opts ...Option) (api.Rep
-	handleHelloMessage := makeHelloHandler(id, n, messageLog, unicastLogs)
+	var handleOwnMessage, handlePeerMessage, handleClientMessage, handleHelloMessage MessageHandler
+	if nonDefaultStack, ok := stack.(NonDefaultMessageHandler); ok {
+		logger.Info("running with non-default message handlers for system testing\n")
```
It seems that we use capitalized style in the existing code consistently (but it does not follow the Go's convention: #105).

```suggestion
		logger.Info("Running with non-default message handlers for system testing")
```

sergefdrv commented on core/replica.go:
```diff
@@ -72,8 +76,14 @@ func New(id uint32, configer api.Configer, stack Stack, opts ...Option) (api.Rep
-	handleOwnMessage, handlePeerMessage, handleClientMessage := defaultMessageHandlers(id, messageLog, unicastLogs, configer, stack, logger)
-	handleHelloMessage := makeHelloHandler(id, n, messageLog, unicastLogs)
+	var handleOwnMessage, handlePeerMessage, handleClientMessage, handleHelloMessage MessageHandler
+	if nonDefaultStack, ok := stack.(NonDefaultMessageHandler); ok {
```
I think there is another way around it. We can rename `New` to `newWithHandlerFunc` and pass a function for generating the handlers as an additional parameter. Then we can define `New` as follows.

```go
func New(id uint32, configer api.Configer, stack Stack, opts ...Option) (api.Replica, error) {
	return newWithHandlerFunc(id, configer, stack, defaultMessageHandlers, opts)
}
```

The testing framework then could use `newWithHandlerFunc` directly.

Alternatively, we can define a new `Option` to supply a function for generating the handlers, but make it unexported/internal.

nhoriguchi replied:
Thanks for suggestion. `newWithHandlerFunc` approach sounds nice to me, so I'll go this direction.

sergefdrv commented on core/bft_test.go:
```diff
@@ -0,0 +1,790 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package minbft_test
```
What are the benefits of placing this test under a different package name?

nhoriguchi replied:
I thought that this new file is the same type of `core/integration_test.go` (not unit tests) so I put it in the same package. Actually I tried at first to put this in package `minbft` so going back to this is possible.

sergefdrv replied:
I am not anymore sure why `core/integration_test.go` belongs to a separate package. Perhaps the original point was that the integration test should not fiddle with the internals of the protocol implementation.

sergefdrv reviewed (commented):
Thank you for sharing this draft. Sorry, I didn't have time to look much into details yet, but I see you made impressive progress with the issue. Let me just share some initial thoughts.

- It looks like the simulator can in principle support any combination of simulated or real replicas, not just a single target replica accompanied by all simulated ones, right?
- Maybe it would be easier to just directly implement scenario-driven replica simulation instead of trying to reuse parts of `message-handling.go`, the message log, etc? I think of scenarios that describe the expected messages in a form of pattern matching and responding with the prescribed messages.
- Wouldn't it be easier to simulate clients in the same manner?
- Isn't it inconvenient to generate/verify real signatures and USIG certificates? Maybe create some mock implementation of the `Authenticator` interface?
- Maybe create a mock implementation of `RequestConsumer` as opposed to using `SimpleLedger`? It must be a bit inconvenient to obtain the expected execution results for the latter one.
- It would be good to specify the operations, that are expected to be executed, directly in the scenario.
- I would suggest to store each scenario in a separate `.yaml` file, under a dedicated [`testdata` subdirectory](https://golang.org/cmd/go/#hdr-Test_packages).

> * the testcase `testManipulatedOperation` seems to reach consensus improperly, where the primary replica manipulates operation string in request message and the manipulated string is stored to `SimpleLedger`. So this might be one example case that one Byzantine faulty node could compromise the operation of MinBFT cluster.

Maybe that is because the operation string is equivalent (`operation: replaced message` vs `request: '{"operation":"replaced message"}'`)?

```yaml
      - type: request
        sender: 0
        operation: replaced message
      - type: prepare
        client: 0
        sender: 0
        receiver: 1
        view: 0
        seq: 0
        prepareid: 0
        request: '{"operation":"replaced message"}'
```

> * I have to update README.md to explain some about Byzantine faulty case.

I think the simulator and its scenario format would benefit from more detailed documentation.

nhoriguchi commented:
> Thank you for sharing this draft. Sorry, I didn't have time to look much into details yet, but I see you made impressive progress with the issue. Let me just share some initial thoughts.

Thank you for the quick feedback.

> * It looks like the simulator can in principle support any combination of simulated or real replicas, not just a single target replica accompanied by all simulated ones, right?

Yes, we can and will support other combinations theoretically, this version just starts with a few simplest cases. Maybe we can have some cases where only faulty nodes are simulated.

> * Maybe it would be easier to just directly implement scenario-driven replica simulation instead of trying to reuse parts of `message-handling.go`, the message log, etc? ...

I started by copying and modifying `core/integration_test.go`, so some (finally unneeded) dependency on internal components of replica still remain. Now I have working code so I can/will refactor the code to be closer to the final form.

> ... I think of scenarios that describe the expected messages in a form of pattern matching and responding with the prescribed messages.

In current version, the definition of expected messages is not flexible enough (and contains some bugs), so I'll try to improve/fix them next.

> * Wouldn't it be easier to simulate clients in the same manner?

Right, I'll find the way for it.

> * Isn't it inconvenient to generate/verify real signatures and USIG certificates? Maybe create some mock implementation of the `Authenticator` interface?
> * Maybe create a mock implementation of `RequestConsumer` as opposed to using `SimpleLedger`? It must be a bit inconvenient to obtain the expected execution results for the latter one.

I'll try to replace them with mocks.

> * It would be good to specify the operations, that are expected to be executed, directly in the scenario.
> * I would suggest to store each scenario in a separate `.yaml` file, under a dedicated [`testdata` subdirectory](https://golang.org/cmd/go/#hdr-Test_packages).

OK, I'll do this.

> > * the testcase `testManipulatedOperation` seems to reach consensus improperly, where the primary replica manipulates operation string in request message and the manipulated string is stored to `SimpleLedger`. So this might be one example case that one Byzantine faulty node could compromise the operation of MinBFT cluster.
> 
> Maybe that is because the operation string is equivalent (`operation: replaced message` vs `request: '{"operation":"replaced message"}'`)?
> 
> ```yaml
>       - type: request
>         sender: 0
>         operation: replaced message

Sorry, this was a wrong input as you point out. But with fixed this (set this "operation" to "test request message"), the block will be generated with "replaced message". I quote the log output of this testcase:

~~~
[minbft] 23:36:11.754 New ▶ INFO 079 Replica 0: running with non-default message handlers for system testing
[minbft] 23:36:11.764 makeMessageStreamHandler.func1 ▶ DEBU 07a Replica 0: Received <HELLO replica=1> from peer
[minbft] 23:36:11.764 makeMessageStreamHandler.func1 ▶ DEBU 07b Replica 0: Handled <HELLO replica=1> from peer
[minbft] 23:36:11.773 New ▶ INFO 07c Replica 2: running with non-default message handlers for system testing
[minbft] 23:36:11.775 makeMessageStreamHandler.func1 ▶ DEBU 07d Replica 1: Received <HELLO replica=0> from peer
[minbft] 23:36:11.775 makeMessageStreamHandler.func1 ▶ DEBU 07e Replica 1: Handled <HELLO replica=0> from peer
[minbft] 23:36:11.776 makeMessageStreamHandler.func1 ▶ DEBU 07f Replica 1: Received <HELLO replica=2> from peer
[minbft] 23:36:11.777 makeMessageStreamHandler.func1 ▶ DEBU 080 Replica 1: Handled <HELLO replica=2> from peer
[minbft] 23:36:11.777 makeMessageStreamHandler.func1 ▶ DEBU 081 Replica 2: Received <HELLO replica=1> from peer
[minbft] 23:36:11.777 makeMessageStreamHandler.func1 ▶ DEBU 082 Replica 2: Handled <HELLO replica=1> from peer
[minbft] 23:36:11.777 makeMessageStreamHandler.func1 ▶ DEBU 083 Replica 0: Received <HELLO replica=2> from peer
[minbft] 23:36:11.778 makeMessageStreamHandler.func1 ▶ DEBU 084 Replica 0: Handled <HELLO replica=2> from peer
[minbft] 23:36:11.778 makeMessageStreamHandler.func1 ▶ DEBU 085 Replica 2: Received <HELLO replica=0> from peer
[minbft] 23:36:11.778 makeMessageStreamHandler.func1 ▶ DEBU 086 Replica 2: Handled <HELLO replica=0> from peer
=== RUN   TestBft/testManipulatedOperation/r=3/c=1/t=1
Step 1/2: send 4 messages
[minbft] 23:36:11.788 makeMessageStreamHandler.func1 ▶ DEBU 087 Replica 2: Received <REQUEST client=0 seq=1602200171745718760 operation="test request message"> from client
[minbft] 23:36:11.788 makeMessageStreamHandler.func1 ▶ DEBU 088 Replica 2: Dropped <REQUEST client=0 seq=1602200171745718760 operation="test request message"> from client
[minbft] 23:36:11.789 makeMessageStreamHandler.func1 ▶ DEBU 089 Replica 0: Received <REQUEST client=0 seq=1602200171745718760 operation="test request message"> from client
[minbft] 23:36:11.789 makeMessageStreamHandler.func1 ▶ DEBU 08a Replica 0: Dropped <REQUEST client=0 seq=1602200171745718760 operation="test request message"> from client
[minbft] 23:36:11.789 makeMessageStreamHandler.func1 ▶ DEBU 08b Replica 1: Received <REQUEST client=0 seq=1602200171745718760 operation="test request message"> from client
replace operation in embedded request with 'replaced message'
[minbft] 23:36:11.798 makeMessageStreamHandler.func1 ▶ DEBU 08c Replica 0: Sending <PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760> to replica 1
[minbft] 23:36:11.798 makeMessageStreamHandler.func1 ▶ DEBU 08d Replica 1: Received <PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760> from replica 0
[minbft] 23:36:11.800 makeGeneratedMessageConsumer.func1 ▶ DEBU 08e Replica 1: Generated <COMMIT cv=1 replica=1 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>>
[minbft] 23:36:11.800 makeMessageStreamHandler.func1 ▶ DEBU 08f Replica 1: Handled <PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760> from replica 0
[minbft] 23:36:11.800 handleOwnPeerMessages ▶ DEBU 090 Replica 1: Handled own <COMMIT cv=1 replica=1 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>>
[minbft] 23:36:11.801 makeMessageStreamHandler.func1 ▶ DEBU 091 Replica 1: Sending <COMMIT cv=1 replica=1 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> to replica 2
[minbft] 23:36:11.801 makeMessageStreamHandler.func1 ▶ DEBU 092 Replica 1: Sending <COMMIT cv=1 replica=1 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> to replica 0
2020/10/08 23:36:11 Received block[1]: replaced message
[minbft] 23:36:11.802 makeGeneratedMessageConsumer.func1 ▶ DEBU 093 Replica 1: Generated <REPLY replica=1 seq=1602200171745718760 result="{\"Height\":1,\"PrevBlockHash\":null,\"Payload\":\"cmVwbGFjZWQgbWVzc2FnZQ==\"}">
[minbft] 23:36:11.802 makeMessageStreamHandler.func1 ▶ DEBU 094 Replica 0: Received <COMMIT cv=1 replica=1 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> from replica 1
[minbft] 23:36:11.802 makeMessageStreamHandler.func1 ▶ DEBU 095 Replica 0: Dropped <COMMIT cv=1 replica=1 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> from replica 1
[minbft] 23:36:11.803 makeMessageStreamHandler.func1 ▶ DEBU 096 Replica 1: Handled <REQUEST client=0 seq=1602200171745718760 operation="test request message"> from client
[minbft] 23:36:11.803 makeMessageStreamHandler.func1 ▶ DEBU 097 Replica 1: Sending <REPLY replica=1 seq=1602200171745718760 result="{\"Height\":1,\"PrevBlockHash\":null,\"Payload\":\"cmVwbGFjZWQgbWVzc2FnZQ==\"}"> to client 0
[minbft] 23:36:11.803 makeMessageStreamHandler.func1 ▶ DEBU 098 Replica 2: Received <COMMIT cv=1 replica=1 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> from replica 1
[minbft] 23:36:11.804 makeMessageStreamHandler.func1 ▶ DEBU 099 Replica 2: Dropped <COMMIT cv=1 replica=1 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> from replica 1
2020/10/08 23:36:11 Received Reply message from replica 1
[minbft] 23:36:11.808 makeMessageStreamHandler.func1 ▶ DEBU 09b Replica 2: Sending <COMMIT cv=1 replica=2 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> to replica 1
[minbft] 23:36:11.809 makeMessageStreamHandler.func1 ▶ DEBU 09c Replica 1: Received <COMMIT cv=1 replica=2 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> from replica 2
[minbft] 23:36:11.810 makeMessageStreamHandler.func1 ▶ DEBU 09d Replica 1: Handled <COMMIT cv=1 replica=2 prepare=<PREPARE cv=1 replica=0 view=0 client=0 seq=1602200171745718760>> from replica 2
[minbft] 23:36:11.819 makeMessageStreamHandler.func1 ▶ DEBU 09e Replica 0: Sending <REPLY replica=0 seq=1602200171745718760 result="{\"Height\":1,\"PrevBlockHash\":null,\"Payload\":\"cmVwbGFjZWQgbWVzc2FnZQ==\"}"> to client 0
2020/10/08 23:36:11 Received Reply message from replica 0
Step 2/2: check 3 messages
====> confirming messages from target replica
3 messages received, 3 expected
FAIL (1/3): msg to 0, expected 2
FAIL (2/3): msg to 2, expected 0
FAIL (3/3): msg result {"Height":1,"PrevBlockHash":null,"Payload":"cmVwbGFjZWQgbWVzc2FnZQ=="}, expected {"Height":1,"PrevBlockHash":null,"Payload":"dGVzdCByZXF1ZXN0IG1lc3NhZ2U="}
=== Simulator done.
--- FAIL: TestBft (0.64s)
    --- PASS: TestBft/testMissingCommit/r=3/c=1/t=1 (0.08s)
    --- PASS: TestBft/testReplayPrepare/r=3/c=1/t=1 (0.10s)
    --- PASS: TestBft/testReplayCommit/r=3/c=1/t=1 (0.10s)
    --- FAIL: TestBft/testManipulatedOperation/r=3/c=1/t=1 (0.10s)
        bft_test.go:956:
                Error Trace:    bft_test.go:956
                Error:          Not equal:
                                expected: 0x0
                                actual  : 0x1
                Test:           TestBft/testManipulatedOperation/r=3/c=1/t=1
~~~
Logs 087-08b show requests from the client, and log 093 shows that replica 1 accepted the modified request. I think that the expected behavior of this case is to detect the forged request and trigger view-change.

> > * I have to update README.md to explain some about Byzantine faulty case.
> 
> I think the simulator and its scenario format would benefit from more detailed documentation.

Yes, I'll add it too. README.md will only contains some simplest case for demonstration. Documentation about simulator itself (with how to add a testcase) will be done in another file (with referring to from README.md).

ynamiki commented on core/bft_test.go:
```diff
@@ -0,0 +1,790 @@
+		targetReplica int
+		testcase      func(t *testing.T)
+	}{
+		{numReplica: 3, numClient: 1, targetReplica: 1, testcase: testSimpleScenarioBackup},
```
How about moving the parameters, `numReplica`, etc. to a YAML as well as a scenario? I am thinking about aggregating a kind of test data and environment definitions to the external structure to separate the code into a test runner and data.

nhoriguchi replied:
Sounds nice, thanks. I'll move the parameters into YAML.

ynamiki reviewed (commented):
Thanks for the patch. I have some small comments.

nhoriguchi committed:
Update core/replica.go

Co-authored-by: Yuta Namiki <yuta.namiki@gmail.com>

sergefdrv commented:
> I started by copying and modifying `core/integration_test.go`, so some (finally unneeded) dependency on internal components of replica still remain. Now I have working code so I can/will refactor the code to be closer to the final form.

What I meant was that we may be better off by fully creating scenario-driven replica/client implementation (i.e. a component that satisfies `api.ConnectionHandler` and uses `api.ReplicaConnector`), without even relying on something like `NonDefaultMessageHandler` or `newWithHandlerFunc`.

> But with fixed this (set this "operation" to "test request message"), the block will be generated with "replaced message". [...] I think that the expected behavior of this case is to detect the forged request and trigger view-change.

Perhaps this happens because the manipulated `Request` message is [properly signed](https://github.com/Naoya-Horiguchi/minbft/blob/087b428f537bdc236a7315d12304328fb5e5162c/core/bft_test.go#L361), therefore not forged. Thus, this case rather represents a Byzantine client generating multiple conflicting but orderly signed requests. In that case, the protocol ensures that all correct replicas execute the same single request of those picked by the primary.

nhoriguchi commented:
> What I meant was that we may be better off by fully creating scenario-driven replica/client implementation (i.e. a component that satisfies `api.ConnectionHandler` and uses `api.ReplicaConnector`), without even relying on something like `NonDefaultMessageHandler` or `newWithHandlerFunc`.

I'll re-implement the framework with such implementations/components.

> 
> > But with fixed this (set this "operation" to "test request message"), the block will be generated with "replaced message". [...] I think that the expected behavior of this case is to detect the forged request and trigger view-change.
> 
> Perhaps this happens because the manipulated `Request` message is [properly signed](https://github.com/Naoya-Horiguchi/minbft/blob/087b428f537bdc236a7315d12304328fb5e5162c/core/bft_test.go#L361), therefore not forged. Thus, this case rather represents a Byzantine client generating multiple conflicting but orderly signed requests. In that case, the protocol ensures that all correct replicas execute the same single request of those picked by the primary.

OK, so this case seems not to be considered as Byzantine replica case (so will be dropped in the next version). I thought that we have to clarify to users/customers the boundary of what the BFT protocol guarantees. So thank you for helping it.

I guess that this test kind of case will be revisited when we'll think of testing BFT of MinBFT with Hyperledger Fabric, where an orderer plays both roles of MinBFT client and MinBFT replica. But the current problem only covers BFT of replicas, not of clients. So I won't mix these separate issues.

nhoriguchi commented:
This PR is left open without update for long, which is not good for community management, so I close this issue for now. I'll create the new PR when the next version gets ready.
